### PR TITLE
mgr/prometheus: fix orch check to prevent Prometheus from crashing 

### PR DESCRIPTION
**Pull request by rkachach**
Fixes: https://tracker.ceph.com/issues/63992

The previous code was calling `get_orch_status` during prometheus module startup. As consequence and since the modules loading order by the mgr is random (sometimes it used alphabetical order, but not always), this can crash in case the orchestrator is not available yet. In summary, the previous code suffers from a race condition between loading prometheus and the orchestrator module so prometheus module could potentially crash during the startup.

Changes purpose is to remove the initialiaztion of the `modify_instance_id` variable from the stratup and move it to the `get_metadata_and_osd_status` which is called both periodically by the metrics collection thread or when some clients performs a get on `/metrics` endpoint. 

Related PR: https://github.com/ceph/ceph/pull/52191
Related issues: https://github.com/rook/rook/issues/13527

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins retest this please

**Review by avanthakkar — APPROVED** [review]
LGTM! The changes worked for me while testing on both orchestrator rook & cephadm. The only issue remaining is the value for `instance_id` label for the rgw metrics where we see mismatch in values of that label exposed by Prometheus module &  ceph-exporter , which I think would involve ceph-exporter changes too. IMO it should be dealt in a seperate PR and having a tracker issue for the same. 
Wdyt? @rkachach 
Based on this I'm approving PR as it solved the issue mentioned, thanks @rkachach

**Comment by rkachach** [comment]
jenkins retest this please

**Review by jmolmo — APPROVED** [review]

**Comment by rkachach** [comment]
jenkins test make check

**Comment by rkachach** [comment]
jenkins test api

File: src/pybind/mgr/prometheus/module.py [hunk 1]
```diff
@@ -1292,18 +1292,19 @@ def _get_pool_info(pool: Dict[str, Any]) -> Tuple[str, str]:
         # Populate other servers metadata
         # If orchestrator is available and ceph-exporter is running modify rgw instance id
         # to match the one from exporter
-        if self.modify_instance_id:
+        modify_instance_id = self.orch_is_available() and self.get_module_option('exclude_perf_counters')
+        if modify_instance_id:
             daemons = raise_if_exception(self.list_daemons(daemon_type='rgw'))
             for daemon in daemons:
+                daemon_id_prefix = str(daemon.daemon_id).split(".")[2] if '.' in daemon.daemon_id else daemon.daemon_id
```
Comment (adk3798) [inline src/pybind/mgr/prometheus/module.py:1299]: mypy failure
```
prometheus/module.py:1299: error: Unsupported right operand type for in ("Optional[str]")
```

File: src/pybind/mgr/prometheus/module.py [hunk 2]
```diff
@@ -1292,18 +1292,19 @@ def _get_pool_info(pool: Dict[str, Any]) -> Tuple[str, str]:
...
                 self.metrics['rgw_metadata'].set(1,
-                                                 ('{}.{}'.format(str(daemon.daemon_type),
-                                                                 str(daemon.daemon_id)),
+                                                 (f"{daemon.daemon_type}.{daemon.daemon_id}",
```
Comment (adk3798) [inline src/pybind/mgr/prometheus/module.py:1301]: mypy failure
```
prometheus/module.py:1301: error: Argument 2 to "set" of "Metric" has incompatible type "Tuple[str, str, str, Optional[str]]"; expected "Optional[Tuple[str, ...]]"
```

**Review by adk3798 — COMMENTED** [review]
I'm mostly okay with this as long as it doesn't bring back the dashboard issue that sparked changes here in the first place. There are a couple related failures in make check that are being caused by this PR though (see review comments)

File: src/pybind/mgr/prometheus/module.py (hunk 1 above)
Comment (rkachach) [reply src/pybind/mgr/prometheus/module.py]: Thanks @adk3798, fixed.

File: src/pybind/mgr/prometheus/module.py (hunk 2 above)
Comment (rkachach) [reply src/pybind/mgr/prometheus/module.py:1304]: fixed

**Comment by rkachach** [comment]
> LGTM! The changes worked for me while testing on both orchestrator rook & cephadm. The only issue remaining is the value for `instance_id` label for the rgw metrics where we see mismatch in values of that label exposed by Prometheus module & ceph-exporter , which I think would involve ceph-exporter changes too. IMO it should be dealt in a seperate PR and having a tracker issue for the same. Wdyt? @rkachach Based on this I'm approving PR as it solved the issue mentioned, thanks @rkachach

I agree on fixing the instance_id issue for RGW in a separate PR. Let's merge and then backport this PR to reef to fix the metrics there as well.

**Pull request merged**